02_knight_and_princess/roman_kuprii/main.cpp

```diff
@@ -0,0 +1,156 @@
+
+#include <array>
+#include <fstream>
+#include <iostream>
+#include <string>
+#include <vector>
+
+using std::endl;
+using std::to_string;
+using std::vector;
+
+using EdgeId = int;
+using VertexId = int;
+
+constexpr int INVALID_ID = -1;
+const std::string JSON_GRAPH_FILENAME = "graph.json";
+
+#define MIN(a, b) ((a) < (b) ? (a) : (b))
+
+struct Edge {
+  const EdgeId id = INVALID_ID;
+  const std::array<VertexId, 2> connected_vertices;
+
+  Edge(const VertexId& start, const VertexId& end, const EdgeId& _id)
+      : id(_id), connected_vertices({start, end}) {}
+
+  std::string to_json() const {
+    std::string res;
+    res = "{ \"id\": ";
+    res += to_string(id);
+    res += ", \"vertex_ids\": [";
+    res += to_string(connected_vertices[0]);
+    res += ", ";
+    res += to_string(connected_vertices[1]);
+    res += "] }";
+    return res;
+  }
+};
+
+struct Vertex {
+  const VertexId id = INVALID_ID;
+  std::vector<EdgeId> edges_ids;
+  int depth = 0;
+
+  explicit Vertex(VertexId _id) : id(_id) {}
+
+  std::string to_json() const {
+    std::string res;
+    res = "{ \"id\": ";
+    res += to_string(id) + ", \"edge_ids\": [";
+    for (int n : edges_ids) {
+      res += to_string(n);
+      res += ", ";
+    }
+    res.pop_back();
+    res.pop_back();
+    res += "], \"depth\": ";
+    res += to_string(depth);
+    res += " }";
+    return res;
+  }
+};
+
+class Graph {
+ public:
+  Graph() {}
+
+  Graph(const vector<Edge>& init_edges, const vector<Vertex>& init_vertices)
+      : vertices_(init_vertices), edges_(init_edges) {}
+
+  std::string to_json() const {
+    std::string res;
+    res = "{ \"vertices\": [ ";
+    for (const auto& v_it : vertices_) {
+      res += v_it.to_json();
+      res += ", ";
+    }
+    res.pop_back();
+    res.pop_back();
+    res += " ], \"edges\": [ ";
+    for (const auto& e_it : edges_) {
+      res += e_it.to_json();
+      res += ", ";
+    }
+    res.pop_back();
+    res.pop_back();
+    res += " ] }\n";
+    return res;
+  }
+
+  void add_vertex() {
+    Vertex new_vertex(vertices_.size());
+    vertices_.push_back(new_vertex);
```

Comment:
Для того чтобы сконструировать новый объект при добавлении в коллекцию, есть специальный метод `emplace`
```suggestion
    vertices_.emplace_back(vertices_.size());
```